dialogue-core/src/main/java/com/palantir/dialogue/core/Responses.java

```diff
@@ -77,5 +86,30 @@ static boolean isClientError(Response response) {
         return response.code() / 100 == 4;
     }
 
+    /**
+     * Returns the integer value of the {@link #PROXY_UPSTREAM_REQUEST_ATTEMPTS} {@link Response#headers() header}.
+     * When the value is missing, negative, or otherwise cannot be parsed, zero is returned.
+     * Rejecting negative values is critical because clients mustn't roll back retry counters based on
+     * responses from a remote server.
+     */
+    static int getProxyUpstreamRequestAttempts(Response response) {
+        Optional<String> maybeProxyUpstreamRequestAttempts = response.getFirstHeader(PROXY_UPSTREAM_REQUEST_ATTEMPTS);
+        if (maybeProxyUpstreamRequestAttempts.isPresent()) {
+            String proxyUpstreamRequestAttempts = maybeProxyUpstreamRequestAttempts.get();
+            try {
+                int parsed = Integer.parseInt(proxyUpstreamRequestAttempts.trim());
+                if (parsed >= 0) {
```

Comment:
Are negative values an indication of something gone wrong? If so, maybe we should we also log a warning when they're encountered as well.

Reply:
Good idea, updated!